infrahub_sdk/store.py

```diff
@@ -20,130 +23,336 @@ def get_schema_name(schema: str | type[SchemaType] | None = None) -> str | None:
+            )
+
+        return node
+
+    def _get_by_key(
+        self, key: str, kind: str | None = None
+    ) -> InfrahubNode | InfrahubNodeSync | CoreNode | CoreNodeSync:
+        if key not in self._keys:
+            raise NodeNotFoundError(
+                node_type="n/a",
+                identifier={"key": [key]},
+                message=f"Unable to find the node {key!r} in the store ({self.branch_name})",
+            )
+
+        node = self._get_by_internal_id(self._keys[key])
+
+        if kind and node.get_kind() != kind:
+            raise NodeInvalidError(
+                node_type=kind,
+                identifier={"key": [key]},
+                message=f"Found a node of kind {node.get_kind()} instead of {kind} for key {key!r} in the store ({self.branch_name})",
+            )
+
+        return node
+
+    def _get_by_id(self, id: str, kind: str | None = None) -> InfrahubNode | InfrahubNodeSync | CoreNode | CoreNodeSync:
+        if id not in self._uuids:
+            raise NodeNotFoundError(
+                node_type="n/a",
+                identifier={"id": [id]},
+                message=f"Unable to find the node {id!r} in the store ({self.branch_name})",
+            )
+
+        node = self._get_by_internal_id(self._uuids[id])
+        if kind and kind not in node.get_all_kinds():
+            raise NodeInvalidError(
+                node_type=kind,
+                identifier={"id": [id]},
+                message=f"Found a node of kind {node.get_kind()} instead of {kind} for id {id!r} in the store ({self.branch_name})",
+            )
+
+        return node
+
+    def _get_by_hfid(
+        self, hfid: str, kind: str | None = None
+    ) -> InfrahubNode | InfrahubNodeSync | CoreNode | CoreNodeSync:
+        if not kind:
+            node_kind, node_hfid = parse_human_friendly_id(hfid)
+        elif kind and hfid.startswith(kind):
+            node_kind, node_hfid = parse_human_friendly_id(hfid)
+        else:
+            node_kind = kind
+            node_hfid = [hfid]
+
+        exception_to_raise_if_not_found = NodeNotFoundError(
+            node_type=node_kind or "unknown",
+            identifier={"hfid": node_hfid},
+            message=f"Unable to find the node {hfid!r} in the store ({self.branch_name})",
+        )
+
+        if node_kind not in self._hfids:
+            raise exception_to_raise_if_not_found
+
+        if tuple(node_hfid) not in self._hfids[node_kind]:
+            raise exception_to_raise_if_not_found
+
+        internal_id = self._hfids[node_kind][tuple(node_hfid)]
+        return self._objs[internal_id]
+
+
+class NodeStoreBase:
+    """Internal Store for InfrahubNode objects.
+
+    Often while creating a lot of new objects,
+    we need to save them in order to reuse them later to associate them with another node for example.
+    """
+
+    def __init__(self, default_branch: str = "main") -> None:
```

Comment:
I think it was like this before but it might be unexpected for users that have configured a different default branch when initializing the SDK and then just use these store objects where we default to "main".

Reply:
I removed the default branch for now but it will introduce a "breaking change" so not 100% sure if we should go with it

Reply:
No with that in mind I'd suggest to keep it as is and instead create an issue so that we cycle back and refactor this in a controlled manner.

Reply:
I added the default branch back along with a deprecation warning